File: Python/2-Medium/00047._全排列_II/00047._全排列_II.py
from typing import List

def permuteUnique(nums: List[int]) -> List[List[int]]:
    def backtrack(start=0):
        # 如果当前排列完成，添加到结果集
        if start == len(nums):
            result.append(nums[:])
            return
        
        lookup = set()
        for i in range(start, len(nums)):
            # 跳过重复元素
            if nums[i] in lookup:
                continue
            lookup.add(nums[i])
            # 交换元素，固定当前第 start 位元素
            nums[start], nums[i] = nums[i], nums[start]
            # 递归填充下一个位置
            backtrack(start + 1)
            # 回溯，撤销操作
            nums[start], nums[i] = nums[i], nums[start]
    
    result = []
    nums.sort()  # 排序，以便去重
    backtrack()
    return result

# 测试代码
test_nums = [1, 1, 2]
print(permuteUnique(test_nums))

File: C++/2-Medium/00047._全排列_II/00047._全排列_II.cpp
#include <vector>
#include <algorithm> // 用于sort函数
#include <iostream>
#include <fstream>
using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> used(nums.size(), false);
        sort(nums.begin(), nums.end()); // 排序，以便于处理重复元素
        backtrack(res, nums, path, used);
        return res;
    }

    void backtrack(vector<vector<int>>& res, vector<int>& nums, vector<int>& path, vector<bool>& used) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (used[i] || (i > 0 && nums[i] == nums[i-1] && !used[i-1])) continue; // 跳过重复元素
            used[i] = true;
            path.push_back(nums[i]);
            backtrack(res, nums, path, used);
            used[i] = false; // 回溯
            path.pop_back();
        }
    }
};

int main() {
    Solution solution;
    vector<int> nums = {1, 1, 2};
    vector<vector<int>> result = solution.permuteUnique(nums);

    // 输出结果
    for (const auto& perm : result) {
        for (int num : perm) {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}